### ISSUE TITLE
Support for bitwise data structures

### ISSUE THREAD

**Issue by psifertex** [Vector35/binaryninja-api#694]
Need support for bitwise data structures.

**Comment by joshwatson** [comment]
Should probably change the title of this to be more explanatory

**Comment by benjaminmordaunt** [comment]
Trying to define a structure like:
```
struct sha_dma_config __packed
{
  union {
    uint32_t cfg;
    struct {
      uint32_t len : 17;
      ...
    }
}
```

... turns into whatever this is:

```
struct sha_dma_config __packed
{
    union
    {
        uint32_t cfg;
        struct
        {
            union
            {
                uint8_t len;
            } bitfield_0;
            union
            {
                uint8_t len;
            } bitfield_1;
            union
            {
                uint8_t len;
                uint8_t irq;
                uint8_t eod;
                uint8_t loop;
                uint8_t mode;
            } bitfield_2;
            union
            {
                uint8_t begin;
                uint8_t end;
                uint8_t op;
                uint8_t unk0;
                uint8_t enconly;
                uint8_t block;
                uint8_t err;
                uint8_t owner;
            } bitfield_3;
        };
    };
};
```

which doesn't appear to be equivalent in any meaningful way. Binary Ninja should really warn that it can't handle this instead of just going along with it and ruining type annotation in my pseudo-C. :-1:

**Comment by vergil-SI** [comment]
Agreed with @benjaminmordaunt. The current solution for this is lackluster, and often more annoying than helpful. Very frustrating to try and deal with bitfield packed structs in Binja currently.

**Comment by bmuessig** [comment]
I agree that it would be great if bitfields could be supported completely and be displayed more clearly

**Comment by AlxCzl** [comment]
Hi! Is there any news concerning this issue? The feature would be quite appreciated as the union workaround isn't very usable.

**Comment by psifertex** [comment]
No updates at this time. Best thing to do is subscribe to the issue to see when we assign it to a release as well as upvote it.

**Comment by utkonos** [comment]
Having support for bitfield types would be very useful. PE relocations are 4 bits and 12 bits, so using a union doesn't cut it.

<img width="453" alt="Image" src="https://github.com/user-attachments/assets/33c4b80f-153f-48a6-87a2-a610aa117d29" />

I was trying to get Sidekick to make and apply the proper type definition, but the dog kept eating Sidekick's homework.

<img width="579" alt="Image" src="https://github.com/user-attachments/assets/f092ccd9-3fdd-4d9b-9b34-38079b587bc6" />

**Comment by Cryptiiiic** [comment]
Just ran into this issue as well
original struct 160 bits(20 bytes)
```c
struct struct2 {
        uint64_t off0:20;
        uint64_t off1:20;
        uint64_t off2:20;
        uint64_t off3:20;
        uint64_t off4:20;
        uint64_t off5:20;
        uint64_t off6:20;
        uint64_t off7:20;
};
```
binja struct 1280 bits(160 bytes)
```c
struct struct2
{
    union
    {
        uint64_t off0;
        uint64_t off1;
        uint64_t off2;
    } __bitfield0;
    union
    {
        uint64_t off3;
        uint64_t off4;
        uint64_t off5;
    } __bitfield8;
    union
    {
        uint64_t off6;
        uint64_t off7;
    } __bitfield10;
};
```

**Comment by TheDuchy** [comment]
Any update on this?

**Comment by psifertex** [comment]
No updates. Whenever we do schedule it for work, it will likely take several releases and we do not have it currently scheduled.

**Comment by pjcreath** [comment]
A possible interim improvement would be to allow comments in types. That way I could at least type the data with the appropriate structure name and length, and readily see my comments about how the fields are actually used, even if BN doesn't truly understand the fields.

**Comment by emesare** [comment]
Support for bitfields in structures has been added in dev builds `>=8462` and will be available for the Io stable (5.2).

The following components also support bitfields: Type Parser (via Clang), Type Printer, SVD Import, PDB Import and DWARF Import

Python docs include an example of defining a bitfield within a structure:
https://dev-docs.binary.ninja/dev/annotation.html#create-bitfields-in-a-structure

Support in data variables:
<img width="415" height="548" alt="Image" src="https://github.com/user-attachments/assets/0c92ecd2-7b19-4e40-a4ce-810d9f371cf8" />

Support in type sidebar:
<img width="482" height="718" alt="Image" src="https://github.com/user-attachments/assets/fe6d186a-9067-4aa8-b85e-ac3f0b367009" />

We do not current do any field resolution for bitfields in HLIL (or otherwise), however that is something we intend on doing in a later release using [type slices](https://github.com/Vector35/binaryninja-api/issues/7512).

Due to our usage of clang you should be able to adjust the bitfield ABI using the appropriate clang flags such as `-mms-bitfields`, when dealing with windows binaries the platform clang target already passes this so the default behavior should be sufficient for most.

**Comment by emesare** [comment]
For analysis to resolve bitfields we have opened the following ticket: https://github.com/Vector35/binaryninja-api/issues/7533 it will likely be prioritized once https://github.com/Vector35/binaryninja-api/issues/7512 is complete.

**Comment by bmuessig** [comment]
That’s great news!
